### PR TITLE
[SPARK-35093] [SQL] AQE now uses newQueryStage plan as key for looking up cached exchanges for re-use

### PR DESCRIPTION
**Pull request by andygrove**
### What changes were proposed in this pull request?
AQE has an optimization where it attempts to reuse compatible exchanges but it does not take into account whether the exchanges are columnar or not, resulting in incorrect reuse under some circumstances.

This PR simply changes the key used to lookup cached stages. It now uses the canonicalized form of the new query stage (potentially created by a plugin) rather than using the canonicalized form of the original exchange.

### Why are the changes needed?
When using the [RAPIDS Accelerator for Apache Spark](https://github.com/NVIDIA/spark-rapids) we sometimes see a new query stage correctly create a row-based exchange and then Spark replaces it with a cached columnar exchange, which is not compatible, and this causes queries to fail.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
The patch has been tested with the query that highlighted this issue. I looked at writing unit tests for this but it would involve implementing a mock columnar exchange in the tests so would be quite a bit of work. If anyone has ideas on other ways to test this I am happy to hear them.

### CODE REVIEW

**Comment by andygrove** [comment]
@tgravescs fyi

**Comment by tgravescs** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @maryannxue FYI

**Comment by dongjoon-hyun** [comment]
Hi, @andygrove . Could you clarify that this is not a correctness issue in the PR description? It only causes job failures, doesn't it?
>  resulting in incorrect reuse under some circumstances.

**Review by viirya — COMMENTED** [review]
Does it mean, if Spark converts reused query stage from columnar to row-based, it will be compatible? Just out of curiosity.

**Comment by andygrove** [comment]
> Hi, @andygrove . Could you clarify that this is not a correctness issue in the PR description? It only causes job failures, doesn't it?

Yes @dongjoon-hyun that is correct. It would result in an invalid plan that would fail to execute.

**Comment by andygrove** [comment]
> Does it mean, if Spark converts reused query stage from columnar to row-based, it will be compatible? Just out of curiosity.

@viirya Spark plugins can create either columnar or row-based exchanges as required for compatibility with other parts of the query plan. 

For example. a columnar `BroadcastHashJoin` would expect both of its child plans to also be columnar, so if Spark replaces one of the child plans with a row-based plan it would not be compatible at execution time.

**Comment by viirya** [comment]
Does `ReuseExchange` rule also possibly suffer from the issue? In the rule, Spark also only checks `canonicalized` plans of exchanges to choose reused exchange.

**Comment by andygrove** [comment]
> Does `ReuseExchange` rule also possibly suffer from the issue? In the rule, Spark also only checks `canonicalized` plans of exchanges to choose reused exchange.

This hasn't been a problem for us. We only ran into issues with the AQE exchange reuse logic because of the adaptive nature where query stages are being created during execution, and with Spark making changes after we provide the new query stage it is too late for our plugin to correct the issue. 

Without AQE, where the `ReuseExchange` rule would be applied (as far as I can see), our plugin would operate on the physical plan after Spark has applied the exchange reuse logic and we would be able to modify the plan as required after that, and before execution, to insert any necessary transitions.

**Comment by andygrove** [comment]
I'll take another look at `ReuseExchange` tomorrow. It might be worth considering adding a similar check here, perhaps as a separate PR.

**Comment by tgravescs** [comment]
If I'm understanding this properly it seems like the root cause is actually that AQE looks at the plan before the columnar changes are applied and physical plan gets executed so it thinks they are canonically the same when really after preparations (specifically here ApplyColumnarRulesAndInsertTransitions) and execution they aren't.  I wonder if there is a better way to do that part.  Might be nice to investigate that a bit further to see if that is possible.

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
> it seems like the root cause is actually that AQE looks at the plan before the columnar changes are applied

This is a good point. Without AQE, `ApplyColumnarRulesAndInsertTransitions` is executed before `ReuseExchange`.

**Comment by tgravescs** [comment]
so I was looking at this today and I think its fairly easy fix to just to change it to store the plan after its been through all the rules.  

`val queryStage = context.stageCache.getOrElseUpdate(newStage.plan.canonicalized, newStage)
`
instead of the 

`val queryStage = context.stageCache.getOrElseUpdate(e.canonicalized, newStage)`

The only downside I can think of is if there is something that is matching on the original plan but after running the rules makes it not canonically the same but really is, it would take a bit longer to have to tranverse down the plan and apply all the rules. But that kind of seems like a bug to me anyway.  
If anyone sees an issue with that let me know, otherwise either @andygrove can update this or I can put up a PR.

**Comment by andygrove** [comment]
Thanks, @tgravescs. That is a much simpler change! I have updated the PR.

**Review by tgravescs — APPROVED** [review]
change looks good to me, but would be great for others to take a look as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 1]
```diff
@@ -431,7 +431,7 @@ case class AdaptiveSparkPlanExec(
               // Check the `stageCache` again for reuse. If a match is found, ditch the new stage
               // and reuse the existing stage found in the `stageCache`, otherwise update the
               // `stageCache` with the new stage.
-              val queryStage = context.stageCache.getOrElseUpdate(e.canonicalized, newStage)
+              val queryStage = context.stageCache.getOrElseUpdate(newStage.plan.canonicalized, newStage)
```
Comment (tgravescs) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:434]: style complaining the line to long, wrap it

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 2]
```diff
@@ -431,7 +431,8 @@ case class AdaptiveSparkPlanExec(
               // Check the `stageCache` again for reuse. If a match is found, ditch the new stage
               // and reuse the existing stage found in the `stageCache`, otherwise update the
               // `stageCache` with the new stage.
-              val queryStage = context.stageCache.getOrElseUpdate(e.canonicalized, newStage)
+              val queryStage = context.stageCache.getOrElseUpdate(
+                newStage.plan.canonicalized, newStage)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:435]: This breaks DPP. I'm not sure exactly why, but the key of stage cache was plan without applying `queryStageOptimizerRules`, now it's different. This probably breaks DPP.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:435]: One idea is to introduce a new phase: `preStageCreationRules`. We can put `ApplyColumnarRulesAndInsertTransitions` and run it before reusing exchanges.

Comment (andygrove) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:435]: Thanks for the suggestions. I am working on this now and will try this out.

Comment (tgravescs) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:435]: thanks for pointing it out. That feels odd to me that it breaks the DPP. I did also notice seems more changes to the way that works in flight: https://github.com/apache/spark/pull/31756/files

Comment (andygrove) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:435]: If we merge https://github.com/apache/spark/pull/31756 and then rebase this PR the DPP tests pass so I think this is the best path forward.

Comment (tgravescs) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:435]: #31756 was merged @andygrove  can you up merge and retest to make sure this works?

**Review by cloud-fan — APPROVED** [review]
LGTM. It will be great to put more details in the PR description, to explain why this simple change fixes the problem.

**Comment by tgravescs** [comment]
@cloud-fan are you ok with pulling this into 3.1.2 and 3.0.3 as well?  those versions wouldn't Support DPP with AQE so shouldn't be any issues.

**Comment by cloud-fan** [comment]
I'm OK with that. The fix is simple and `supportColumns` API is in 3.0/3.1

**Comment by tgravescs** [comment]
@andygrove could update the description and then test on the 3.1.2 and 3.0.3 branches when you get a chance?

**Comment by andygrove** [comment]
@cloud-fan @tgravescs I updated the title and description. Let me know if this is still not clear.

**Review by tgravescs — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Yes, please backport this to the old branches.

**Comment by andygrove** [comment]
I applied this patch to the latest in branch-3.0 and branch-3.1 and ran manual tests to confirm that this fixes the issue for us in those releases.

**Comment by tgravescs** [comment]
I'm not sure why the label pull requests check isn't running here.  Anyone know?

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Not only here, I also saw other PRs their label pull requests are queued too.

**Comment by dongjoon-hyun** [comment]
Hi, All. This seems to break `branch-3.0`.
- https://github.com/apache/spark/runs/2621206825

**Comment by dongjoon-hyun** [comment]
Could you check if that is a flakiness or not?
```
[info] AdaptiveQueryExecSuite:
30483
[info] - Change merge join to broadcast join (521 milliseconds)
30484
[info] - Reuse the parallelism of CoalescedShuffleReaderExec in LocalShuffleReaderExec (265 milliseconds)
30485
[info] - Reuse the default parallelism in LocalShuffleReaderExec (260 milliseconds)
30486
[info] - Scalar subquery (422 milliseconds)
30487
[info] - Scalar subquery in later stages (567 milliseconds)
30488
[info] - multiple joins *** FAILED *** (1 second, 25 milliseconds)
30489
[info]   ArrayBuffer(BroadcastHashJoin [b#147752], [a#147761], Inner, BuildLeft
```

**Comment by andygrove** [comment]
I'm looking now

**Comment by andygrove** [comment]
The tests worked fine for me locally at commit 706f91e5e0ae3c6aa156232546f815840f977201. AQE is non-deterministic because query stages can change depending on the order of other stages completing, so this is likely a flakey test. I will investigate further and see how we can make it more robust.

**Comment by andygrove** [comment]
It looks like this is not a new issue, although potentially the recent change made it more likely to happen. We should probably re-open https://issues.apache.org/jira/browse/SPARK-32304

**Comment by HyukjinKwon** [comment]
Late LGTM2

**Pull request closed without merging**